### ISSUE TITLE
Refactor: Extract Function

### ISSUE THREAD

**Issue by markmandel** [cursive-ide/cursive#89]
The ability to highlight an s-expression and extract a new function out from it in the same module, much like you can with Ruby/JS/Java etc.

**Comment by arnelandwehr** [comment]
+1

**Comment by DjebbZ** [comment]
I have to +1 the feature too.

**Comment by kurman** [comment]
+1. I tried using "extract variable" option and then move it out manually. However it is error prone when symbols are bound in outer scope.

**Comment by jitlogic** [comment]
+1

**Comment by jitlogic** [comment]
Also, 'inline function' feature would be useful so it can work in both ways. Just as Extract Method and Inline Method in IntelliJ Java refactoring tools.

**Comment by hellonico** [comment]
back to using intellij as main clojure dev env, and ... if any help to test ... drop me a line.

**Comment by verma** [comment]
+1, yes please.

**Comment by marcolisi** [comment]
+1, yes please.

**Comment by dinuksw** [comment]
+ 1, yes please

**Comment by ruttenberg** [comment]
+1, please

**Comment by smartoma** [comment]
+1, please

**Comment by winechickc** [comment]
+1

**Comment by aadrian** [comment]
+1

**Comment by DjebbZ** [comment]
Sorry to bump this issue again. I would really love this kind of refactoring to be available in Cursive, as well as being able to extract anything to the top level scope of the namespace. I'm a huge "refactorer".

Idea :
- highlight an s-expression
- execute the Extract command
- in the scope selection, have "top-level"
- select the kind of form : `def`, `defn`, `defmacro`, and ideally every stuff marked as `def`/`defn`/`defmacro` in the Cursive options.
- extract & profit !

**Comment by yilazius** [comment]
Is this feature under consideration at all? 
It is pretty much essential for heaving an IDE driven refactoring workflow IMHO.
Work on this would be greatly appreciated!
Cheers!

**Comment by deepakazad** [comment]
+1

**Comment by myeh98** [comment]
+1

**Comment by IvanVas** [comment]
The feature would be very useful.

**Comment by Stvad** [comment]
This is an essential refactoring functionality, unpleasantly surprised that this has been ignored for years :(   
Do people know any workarounds?

**Comment by khangharoth** [comment]
Any reason it's not been picked or prioritized?

**Comment by ericauld** [comment]
Eager for this feature

**Comment by beatrizanami** [comment]
This would be amazing :)

**Comment by esuomi** [comment]
I'm joining with the crowd above - extracting new functions from existing code is _essential_ for any IDE, and it pairs really well with duplicate code elimination.

**Comment by JoshuaFox** [comment]
Yes please. And given Clojure's simple structure, this would seem to be even easier than for other languages.

**Comment by ionutbilica** [comment]
+1

**Comment by joseslneto** [comment]
we forgot to celebrate the 10 year anniversary of this feature 😅
